File: AVL.cpp
#include <bits/stdc++.h>

using namespace std;

class AVLTree
{
private:

    struct Node
    {
        int v;
        Node* l = nullptr;
        Node* r = nullptr;
        int h = 1;
        int s = 1;

        Node(int v)
        {
            this->v = v;
        }

        int balance()
        {
            return (r ? r->h : 0) - (l ? l->h : 0);
        }

        void fix()
        {
            h = max(r ? r->h : 0, l ? l->h : 0) + 1;
            s = (r ? r->s : 0) + (l ? l->s : 0) + 1;
        }
    };

    Node* head = nullptr;
    int size = 0;

    function<Node*(Node*)> _min = [&](Node* n) { return n->l ? _min(n->l) : n; };
    function<Node*(Node*)> _max = [&](Node* n) { return n->r ? _max(n->r) : n; };

    Node* small_rotate(Node* node, bool right)
    {
        Node* t;
        if (right)
        {
            t = node->l;
            node->l = t->r;
            t->r = node;
        }
        else
        {
            t = node->r;
            node->r = t->l;
            t->l = node;
        }

        node->fix();
        t->fix();

        return t;
    }

    Node* balance(Node* node)
    {
        node->fix();
        if (node->balance() == -2)
        {
            if (node->l->balance() == 1)
                node->l = small_rotate(node->l, false);
            return small_rotate(node, true);
        }

        if (node->balance() == 2)
        {
            if (node->r->balance() == -1)
                node->r = small_rotate(node->r, true);
            return small_rotate(node, false);
        }

        return node;
    }

public:

    void add(int v)
    {
        ++size;
        function<Node*(Node*)> deep_add = [&](Node* node)
        {
            if (!node)
                return new Node(v);
            if (v >= node->v)
                node->r = deep_add(node->r);
            else
                node->l = deep_add(node->l);
            return balance(node);
        };
        head = deep_add(head);
    }

    void del(int v)
    {
        --size;
        function<Node*(Node*)> deep_del = [&](Node* node)
        {
            if (v > node->v)
                node->r = deep_del(node->r);
            else if (v < node->v)
                node->l = deep_del(node->l);
            else
            {
                Node* l = node->l;
                Node* r = node->r;
                delete node;
                if (!r)
                    return l;

                function<Node*(Node*)> del_min = [&](Node* n)
                {
                    if (!n->l)
                        return n->r;

                    n->l = del_min(n->l);
                    return balance(n);
                };
                Node* t = _min(r);
                t->r = del_min(r);
                t->l = l;
                return balance(t);
            }

            return balance(node);
        };
        head = deep_del(head);
    }

    int get_size() { return size; }

    int get_max() { return _max(head)->v;}
    int get_min() { return _min(head)->v;}

    bool check(Node* node)
    {
        if (node)
        {
            if (!check(node->l) || !check(node->r))
                return false;
            node->fix();
            return abs(node->balance()) < 2;
        }

        return true;
    }

    int nth_element(int k)
    {
        function<int(Node*, int)> dfs = [&](Node* node, int n)
        {
            int l = node->l ? node->l->s : 0;
            if (n == l)
                return node->v;
            if (n > l)
                return dfs(node->r, n - l - 1);
            return dfs(node->l, n);
        };

        return dfs(head, k);
    }

    pair<bool, int> lower_bound(int v)
    {
        if (v > get_max() || v < get_min())
            return {false, 0};

        auto less = [&](Node* a, Node* b)
        {
            if (!a || a->v < v)
                return b;
            if (!b || b->v < v)
                return a;
            return a->v < b->v ? a : b;
        };

        function<Node*(Node*)> lb = [&](Node* node)
        {
            return node ? less(node, lb(v > node->v ? node->r : node->l)) : node;
        };

        return {true, lb(head)->v};
    }

    ~AVLTree()
    {
        function<void(Node*)> destroy = [&](Node* node)
        {
            if (node)
            {
                destroy(node->l);
                destroy(node->r);
                delete node;
            }
        };
        destroy(head);
    }

};

int main()
{
    AVLTree tree;
    
    for (int i=0; i < 10; ++i)
        tree.add(5 * i);
    
    cout << tree.lower_bound(11).second << '\n';
    
    return 0;
}

File: AVL.py
#! /usr/bin/env python

class Node:
	def __init__(self, v):
		self.v = v
		self.l, self.r = None, None
		self.h, self.size = 1, 1

	def fix(self):
		self.h = max(self.r.h if self.r else 0, self.l.h if self.l else 0) + 1
		self.size = (self.r.size if self.r else 0) + (self.l.size if self.l else 0) + 1

	def balance(self):
		return (self.r.h if self.r else 0) - (self.l.h if self.l else 0)

class AVLTree:
	def rotate(node, is_right):
		if is_right:
			t = node.l
			node.l = t.r
			t.r = node
		else:
			t = node.r
			node.r = t.l
			t.l = node
		
		node.fix()
		t.fix()

		return t

	def balance(node):
		node.fix()
		if node.balance() == -2:
			if node.l.balance() == 1:
				node.l = AVLTree.rotate(node.l, False)
			return AVLTree.rotate(node, True)
		if node.balance() == 2:
			if node.r.balance() == -1:
				node.r = AVLTree.rotate(node.r, True)
			return AVLTree.rotate(node, False)
		return node

	def get_min(node):
		return AVLTree.get_min(node.l) if node.l else node

	def get_max(node):
		return AVLTree.get_max(node.r) if node.r else node

	def __init__(self):
		self.size = 0
		self.head = None


	def add(self, v):
		self.size += 1

		def Add(node):
			if not node:
				return Node(v)
			t = node.r if v >= node.v else node.l
			if v >= node.v:
				node.r = Add(node.r)
			else:
				node.l = Add(node.l)
			return AVLTree.balance(node)

		self.head = Add(self.head)


	def delete(self, v):
		self.size -= 1
		def Delete(node):
			if v > node.v:
				node.r = Delete(node.r)
			elif v < node.v:
				node.l = Delete(node.l)
			else:
				l, r = node.l, node.r
				del node
				if not r:
					return l
				
				def del_min(n):
					if not n.l:
						return n.r
					n.l = del_min(n.l)
					return balance(n)
				t = AVLTree.get_min(r)
				t.r = del_min(r)
				t.l = l
				return AVLTree.balance(t)
			return AVLTree.balance(node)
		self.head = Delete(self.head)


	def max(self):
		return AVLTree.get_max(self.head).v

	def min(self):
		return AVLTree.get_min(self.head).v

	def nth_element(self, n):
		def Nth_element(node, k):
			l = node.l.size if node.l else 0
			if k == l:
				return node.v
			if k > l:
				return Nth_element(node.r, k - l - 1)
			return Nth_element(node.l, k)
		return Nth_element(self.head, n)

	def print(self):
		def Print(node):
			if node:
				Print(node.l)
				Print(node.r)
				print(node.v, node.h)
		Print(self.head)


if __name__ == '__main__':
	tree = AVLTree()
	for i in range(10):
		tree.add(i)
	tree.print();
	# print(tree.nth_element(64))
	# print(tree.min(), tree.max())
	# for i in range(15):
	# 	tree.delete(i)
	# print(tree.min(), tree.max())
	# for i in range(99, 80, -1):
	# 	tree.delete(i)
	# print(tree.min(), tree.max())
	# print(tree.nth_element(64))